### PR TITLE
Bluetooth: Mesh: Add OOB Public Key support for provisionee role

### PR DESCRIPTION
**Pull request by PavelVPV**
This commit allows an unprovisioned device to exchange its public key using
out-of-band techology. See MeshPRFv1.0.1, table 5.19 and section 5.4.2.3.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

### CODE REVIEW

**Comment by PavelVPV** [comment]
Mesh uses controller to generate DH key. I guess, to support static OOB public key, DH key generation should be moved up to mesh.

**Comment by PavelVPV** [comment]
Clarification on how it works (perhaps I should make this clear in API). The key pair can be either dynamic or static.

To use dynamic key:
1. Set `pub_key_set` to `BT_MESH_PUB_KEY_OOB`;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L115
2. Call `bt_mesh_prov_local_pub_key_cb_register` to retrieve generated public key before provisioning starts to send it OOB;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L304

Dynamic pair re-generated on every provisioning attempt (as without OOB Public Key). The user callback will be called every time when a new public key is generated. Perhaps, this won't happen, if the key pair is re-generated implicitly by BLE stack before the provisioning happens.

To use static key pair:
1. Set `pub_key_set` to `BT_MESH_PUB_KEY_OOB`;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L115
2. Initialise `public_key` and `private_key` pointers before provisioning is started;https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/include/bluetooth/mesh/main.h#L117-L120

In case of static key pair, DH key generated directly in provisionee module instead of using controller: https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L149-L150
https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L346-L348
https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L312-L338

File: include/bluetooth/mesh/main.h [hunk 1]
```diff
@@ -105,6 +111,14 @@ struct bt_mesh_prov {
 	/** Out of Band information field. */
 	bt_mesh_prov_oob_info_t oob_info;
 
+	/** Public Key type */
+	enum bt_mesh_pub_key_type_t pub_key_type;
+
+	/** Pointer to Public Key for static OOB Public Key */
+	const uint8_t *public_key;
+	/** Pointer to Private Key for static OOB Public Key */
+	const uint8_t *private_key;
```
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:120]: This is different from the provisioner side of the OOB mechanism, which uses a function. It would be nice to use the same pattern (if not the same mechanism), but I don't see any obvious solution to this. Any suggestions?

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
Is there a usecase that we need to support with the dynamic OOB key, or could we just stick with the static OOB key for those, and require users to set them beforehand? AFAICT, it's possible to achieve the same things with this approach as well, although it moves some extra work into the app.

This could remove the need for the `pub_key_type` field and the pubkey callback, and would make the mixed dh_key_gen mechanism more clear cut and easier to follow. The key generation part was already the most intricate of the provisioning state machine, so it'd be nice if we can avoid adding another dimension to it.

I agree that this needs some conceptual documentation, even if we remove the dynamic OOB key mechanism. E.g.:
- When to use OOB public key
- Requirements for the key material
- A short snippet showing how, since it's not used in any samples

File: subsys/bluetooth/mesh/prov_device.c [hunk 2]
```diff
@@ -311,7 +368,24 @@ static void prov_pub_key(const uint8_t *data)
...
 	memcpy(&bt_mesh_prov_link.conf_inputs[17], data, 64);
 
-	if (!bt_pub_key_get()) {
+	if (atomic_test_bit(bt_mesh_prov_link.flags, OOB_PUB_KEY)) {
+		const uint8_t *local_pk;
+
+		if (atomic_test_bit(bt_mesh_prov_link.flags, OOB_PUB_KEY_STATIC)) {
+			local_pk = bt_mesh_prov->public_key;
+		} else {
+			/* If we got to this point, Public Key is already generated. */
```
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/prov_device.c:377]: When OOB Public key type is chosen by provisioner, it (by spec, figure 5.15) sends its public key after it reads out provisionee's public key over OOB. Although, provisionee should probably fail here if public key is not ready.

File: include/bluetooth/mesh/main.h (hunk 1 above)
Comment (PavelVPV) [reply include/bluetooth/mesh/main.h:120]: I guess in case of provisioner the idea is that public key changes from provisionee to provisionee, while for provisionee it's kind of static, which doesn't make sense to use a function.

Comment (trond-snekvik) [reply include/bluetooth/mesh/main.h:120]: That's true, yeah. I might be looking for a pattern where there is none :slightly_smiling_face:

File: subsys/bluetooth/mesh/provisioner.c [hunk 3]
```diff
@@ -724,8 +717,8 @@ int bt_mesh_prov_remote_pub_key_set(const uint8_t public_key[64])
 	}
 
 	/* Swap X and Y halves independently to big-endian */
-	memcpy(&bt_mesh_prov_link.conf_inputs[81], public_key, 32);
-	memcpy(&bt_mesh_prov_link.conf_inputs[81 + 32], &public_key[32], 32);
+	sys_memcpy_swap(&bt_mesh_prov_link.conf_inputs[81], public_key, 32);
+	sys_memcpy_swap(&bt_mesh_prov_link.conf_inputs[81 + 32], &public_key[32], 32);
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/provisioner.c:721]: Can you put this in a separate commit?

Comment (LingaoM) [reply subsys/bluetooth/mesh/provisioner.c:721]: https://github.com/zephyrproject-rtos/zephyr/blob/efccfcbfc91501cac2ede0541a48c338f2b0cae4/subsys/bluetooth/mesh/provisioner.c#L414-L415

Have done the same thing here?

File: subsys/bluetooth/mesh/prov_device.c [hunk 4]
```diff
@@ -280,9 +306,35 @@ static void prov_dh_key_cb(const uint8_t dhkey[32])
...
+	sys_memcpy_swap(private_key_be, bt_mesh_prov->private_key, 32);
+
+	ret = uECC_valid_public_key(remote_pk_be, &curve_secp256r1);
+	if (ret < 0) {
+		BT_ERR("public key is not valid (ret %d)", ret);
+		ret = TC_CRYPTO_FAIL;
+	} else {
+		ret = uECC_shared_secret(remote_pk_be, private_key_be,
+					 bt_mesh_prov_link.dhkey, &curve_secp256r1);
+	}
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/prov_device.c:329]: If we put ECC in Mesh, do we need to re-evaluate Stack resource occupation?
https://github.com/zephyrproject-rtos/zephyr/blob/efccfcbfc91501cac2ede0541a48c338f2b0cae4/subsys/bluetooth/mesh/Kconfig#L261

File: subsys/bluetooth/mesh/prov_device.c [hunk 5]
```diff
@@ -291,16 +343,21 @@ static void prov_dh_key_gen(void)
...
+		static_dh_key_gen();
+	} else {
+		/* Copy remote key in little-endian for bt_dh_key_gen().
+		 * X and Y halves are swapped independently. The bt_dh_key_gen()
+		 * will also take care of validating the remote public key.
+		 */
+		sys_memcpy_swap(remote_pk_le, remote_pk, 32);
+		sys_memcpy_swap(&remote_pk_le[32], &remote_pk[32], 32);
+
+		if (bt_dh_key_gen(remote_pk_le, prov_dh_key_cb)) {
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/prov_device.c:357]: Is it necessary to use the existing structure and add specific functions so that the bt controller can calculate the given private/public key ?

File: subsys/bluetooth/mesh/prov_device.c [hunk 6]
```diff
@@ -280,9 +306,35 @@ static void prov_dh_key_cb(const uint8_t dhkey[32])
...
+	if (ret < 0) {
+		BT_ERR("public key is not valid (ret %d)", ret);
+		ret = TC_CRYPTO_FAIL;
+	} else {
+		ret = uECC_shared_secret(remote_pk_be, private_key_be,
+					 bt_mesh_prov_link.dhkey, &curve_secp256r1);
+	}
 
-	send_pub_key();
+	if (ret == TC_CRYPTO_FAIL) {
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/prov_device.c:331]:
```suggestion
	if (ret != TC_CRYPTO_SUCCESS) {
```

File: include/bluetooth/mesh/main.h [hunk 7]
```diff
@@ -272,6 +286,23 @@ int bt_mesh_input_string(const char *str);
...
 
+/** @brief Register a callback to retrieve a local public key.
+ *
+ * This function needs to be used when @ref BT_MESH_PUB_KEY_OOB is set.
+ * The retrieved public key has to be shared with a Provisioner using
+ * any available OOB technology.
+ *
+ * A new key is generated after each provisioning attempt.
+ *
+ * Used only by Provisionee.
```
Comment (LingaoM) [inline include/bluetooth/mesh/main.h:297]: According to https://github.com/zephyrproject-rtos/zephyr/pull/28457#discussion_r501068834 .
Should it be:
```suggestion
 * Used only by device provisioning.
```

File: include/bluetooth/mesh/main.h [hunk 8]
```diff
@@ -272,6 +286,23 @@ int bt_mesh_input_string(const char *str);
...
+ *
+ * @param cb Callback that will be called when the public key is ready.
+ * Returns a local 64-bytes long public key generated for the provisioning.
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+int bt_mesh_prov_local_pub_key_cb_register(void (*cb)(const uint8_t *key));
```
Comment (LingaoM) [inline include/bluetooth/mesh/main.h:304]: Why implement such a function?

This function only provides public key capabilities, and does not have private key information,
so even if the App has a public key, it's still cannot use it for `prov->private_key`. 

https://github.com/zephyrproject-rtos/zephyr/blob/fa57145f180e1249948b273c7e2dc006ce53f21e/subsys/bluetooth/mesh/prov_device.c#L320

So public/private key can only be static.

Comment (LingaoM) [reply include/bluetooth/mesh/main.h:304]: My suggestion is that the App layer defines its own way of generating public and private keys. The protocol stack does not provide capabilities. Therefore, this registration can be omitted.

File: subsys/bluetooth/mesh/prov_device.c [hunk 9]
```diff
@@ -13,6 +13,11 @@
 #include <sys/util.h>
 #include <sys/byteorder.h>
 
+#include <tinycrypt/constants.h>
+#include <tinycrypt/utils.h>
+#include <tinycrypt/ecc.h>
+#include <tinycrypt/ecc_dh.h>
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/prov_device.c:19]: https://github.com/zephyrproject-rtos/zephyr/blob/38eaabc9c75c9a0932bbd272952dc81a2f2939ef/subsys/bluetooth/mesh/Kconfig#L261-L266

If we want to implement ECC in Mesh, do we need to re-evaluate the resource occupation of ADV Stack?

[the remark this answers is not in the document]
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/prov_device.c:19]: I don't think so, this does not add calls to the public key generation from the send callbacks

[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/prov_device.c:19]: `prov_pub_key` --> `prov_dh_key_gen` --> `static_dh_key_gen`

[the remark this answers is not in the document]
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/prov_device.c:19]: `prov_pub_key` comes from the scanner thread though

[the remark this answers is not in the document]
Comment (LingaoM) [reply subsys/bluetooth/mesh/prov_device.c:19]: In fact, should it be `BT_RX_STACK_SIZE` ?

File: subsys/bluetooth/mesh/prov_device.c [hunk 10]
```diff
@@ -280,9 +306,35 @@ static void prov_dh_key_cb(const uint8_t dhkey[32])
 
 	sys_memcpy_swap(bt_mesh_prov_link.dhkey, dhkey, 32);
 
-	BT_DBG("DHkey: %s", bt_hex(bt_mesh_prov_link.dhkey, 32));
+	dh_key_gen_complete();
+}
+
+static void static_dh_key_gen(void)
```
Comment (alxelax) [inline subsys/bluetooth/mesh/prov_device.c:312]: Shouldn't it be better to keep all ECC dependent algorithms in crypto.c\.h?

File: subsys/bluetooth/mesh/prov_device.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/prov_device.c:19]: Thread usage is the same in both cases: with and without OOB public key:
```
Thread analyze:         
 BT Mesh adv         : STACK: unused 248 usage 520 / 768 (67 %); CPU: 0 %
 BT RX               : STACK: unused 864 usage 1184 / 2048 (57 %); CPU: 2 %
 BT RX pri           : STACK: unused 296 usage 152 / 448 (33 %); CPU: 0 %
 BT ECC              : STACK: unused 256 usage 888 / 1144 (77 %); CPU: 1 %
 BT TX               : STACK: unused 272 usage 368 / 640 (57 %); CPU: 0 %
 sysworkq            : STACK: unused 7048 usage 1144 / 8192 (13 %); CPU: 0 %
 shell_uart          : STACK: unused 1616 usage 432 / 2048 (21 %); CPU: 0 %
 idle 00             : STACK: unused 264 usage 56 / 320 (17 %); CPU: 95 %
```

**Comment by PavelVPV** [comment]
Thanks everyone for the comments! I hope they are resolved now.

**Review by LingaoM — APPROVED** [review]
Great work. LGTM!

File: subsys/bluetooth/mesh/prov_device.c [hunk 11]
```diff
@@ -92,8 +97,8 @@ static void prov_invite(const uint8_t *data)
 	/* Supported algorithms - FIPS P-256 Eliptic Curve */
 	net_buf_simple_add_be16(&buf, BIT(PROV_ALG_P256));
 
-	/* Public Key Type, Only "No OOB" Public Key is supported */
-	net_buf_simple_add_u8(&buf, PUB_KEY_NO_OOB);
+	/* Public Key Type */
+	net_buf_simple_add_u8(&buf, !!bt_mesh_prov->public_key_be);
```
Comment (jhedberg) [inline subsys/bluetooth/mesh/prov_device.c:101]: I know it'll be more lines of code, but I think it would be better to explicitly map to one of the allowed Public Key type values to make it clear that you're translating from some internal implementation detail (NULL pointer or not) to a valid provisioning protocol value.

So something like:
```
if (bt_mesh_prov->public_key_be == NULL) {
        net_buf_simple_add_u8(buf, PUB_KEY_NO_OOB);
} else {
        net_buf_simple_add_u8(buf, PUB_KEY_OOB);
}

File: subsys/bluetooth/mesh/prov_device.c (hunk 10 above)
Comment (PavelVPV) [reply subsys/bluetooth/mesh/prov_device.c]: It could be moved, but would it make a difference?

Comment (alxelax) [reply subsys/bluetooth/mesh/prov_device.c]: at the current moment no. However, if someone :) decides to use another security toolbox or to move this code to the crypto memory area it would be done easier.

File: include/bluetooth/mesh/main.h [hunk 12]
```diff
@@ -105,6 +105,23 @@ struct bt_mesh_prov {
...
+	 * Remember to enable :c:option:`CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY`
+	 * when initializing this parameter.
+	 *
+	 * Must be used together with :c:member:`struct bt_mesh_prov.private_key_be.
+	 */
+	const uint8_t *public_key_be;
+	/** Pointer to Private Key in big-endian for OOB public key type support.
+	 *
+	 * Remember to enable :c:option:`CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY`
+	 * when initializing this parameter.
```
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:119]: I'd suggest adding a warning in `bt_mesh_prov_enable` if a keypair is set, but OOB pubkey is disabled. E.g.
```c
	if (!IS_ENABLED(CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY) && bt_mesh_prov->public_key_be) {
		BT_WARN("OOB Public key disabled");
	}
```

File: include/bluetooth/mesh/main.h [hunk 13]
```diff
@@ -105,6 +105,23 @@ struct bt_mesh_prov {
...
+	 *
+	 * Must be used together with :c:member:`struct bt_mesh_prov.public_key_be.`
```
Comment (trond-snekvik) [inline include/bluetooth/mesh/main.h:121]: Perhaps we could return an error from `bt_mesh_prov_enable` if only one is non-null? E.g.
```c
	if (!!bt_mesh_prov->public_key_be != !!bt_mesh_prov->private_key_be) {
		BT_WARN("Both public and private key must be set");
		return -EINVAL;
	}
```

File: include/bluetooth/mesh/main.h (hunk 12 above)
Comment (PavelVPV) [reply include/bluetooth/mesh/main.h:119]: I was thinking about that, but I decided that one warning in `prov_device.c` (https://github.com/zephyrproject-rtos/zephyr/blob/04e81153e066577d39a96a209d9444182cd03aa3/subsys/bluetooth/mesh/prov_device.c#L145) is enough. This is something that you do once since it's compile time option.

File: include/bluetooth/mesh/main.h (hunk 13 above)
Comment (PavelVPV) [reply include/bluetooth/mesh/main.h:121]: There is the warning here: https://github.com/zephyrproject-rtos/zephyr/blob/04e81153e066577d39a96a209d9444182cd03aa3/subsys/bluetooth/mesh/prov_device.c#L145

File: include/bluetooth/mesh/main.h (hunk 12 above)
Comment (trond-snekvik) [reply include/bluetooth/mesh/main.h:119]: Alright, sounds fair enough :+1:

File: include/bluetooth/mesh/main.h [hunk 14]
```diff
@@ -105,6 +105,23 @@ struct bt_mesh_prov {
 	/** Out of Band information field. */
 	bt_mesh_prov_oob_info_t oob_info;
 
+	/** Pointer to Public Key in big-endian for OOB public key type support.
+	 *
+	 * Remember to enable :c:option:`CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY`
+	 * when initializing this parameter.
+	 *
+	 * Must be used together with :c:member:`struct bt_mesh_prov.private_key_be.
```
[the remark this answers is not in the document]
Comment (trond-snekvik) [reply include/bluetooth/mesh/main.h]: I'm not 100% sure, but I believe this doesn't work with `struct blahblah`, it needs to just be `blahblah`

**Review by trond-snekvik — APPROVED** [review]
